src/Razor/src/Microsoft.AspNetCore.Razor.Language/Legacy/HtmlMarkupParser.cs

```diff
@@ -233,10 +236,391 @@ private SyntaxList<RazorSyntaxNode> ParseTagBody(string tagName, out bool seenEn
+                        while (!EndOfFile && !IsEndOfAttributeValue(quote, CurrentToken))
+                        {
+                            ParseAttributeValue(attributeValueBuilder, quote);
+                        }
+
+                        if (attributeValueBuilder.Count > 0)
+                        {
+                            attributeValue = SyntaxFactory.GenericBlock(attributeValueBuilder.ToList());
+                        }
+                    }
+                }
+
+                // Capture the suffix
+                if (quote != SyntaxKind.Marker && At(quote))
+                {
+                    AcceptAndMoveNext();
+                    // Again, block chunk generator will render the suffix
+                    SpanContext.ChunkGenerator = SpanChunkGenerator.Null;
+                    valueSuffix = OutputAsMarkupLiteral();
+                }
+            }
+            else if (quote != SyntaxKind.Marker || !whitespaceAfterEquals.Any())
+            {
+                valuePrefix = OutputAsMarkupLiteral();
+
+                using (var pooledResult = Pool.Allocate<RazorSyntaxNode>())
+                {
+                    var attributeValueBuilder = pooledResult.Builder;
+                    // Not a "conditional" attribute, so just read the value
+                    ParseMarkupNodes(attributeValueBuilder, ParseMode.Text, token => IsEndOfAttributeValue(quote, token));
+
+                    // Output already accepted tokens if any as markup literal
+                    var literalValue = OutputAsMarkupLiteral();
+                    attributeValueBuilder.Add(literalValue);
+
+                    // Capture the attribute value (will include everything in-between the attribute's quotes).
+                    attributeValue = SyntaxFactory.GenericBlock(attributeValueBuilder.ToList());
+                }
+
+                if (quote != SyntaxKind.Marker)
+                {
+                    TryAccept(quote);
+                    valueSuffix = OutputAsMarkupLiteral();
+                }
+            }
+            else
+            {
+                // There is no quote and there is whitespace after equals. There is no attribute value.
+            }
+
+            return SyntaxFactory.MarkupAttributeBlock(namePrefix, name, nameSuffix, equalsToken, valuePrefix, attributeValue, valueSuffix);
+        }
+
+        private void ParseAttributeValue(in SyntaxListBuilder<RazorSyntaxNode> builder, SyntaxKind quote)
         {
-            // Calling to legacy code for now.
-            LegacyParseTagContent(builder);
+            var prefixStart = CurrentStart;
+            var prefixTokens = ReadWhile(token => token.Kind == SyntaxKind.Whitespace || token.Kind == SyntaxKind.NewLine);
+
+            if (At(SyntaxKind.Transition))
+            {
+                if (NextIs(SyntaxKind.Transition))
+                {
+                    // Wrapping this in a block so that the ConditionalAttributeCollapser doesn't rewrite it.
```

Comment:
This is done at a different layer right? Does this still apply?

Reply:
The comment doesn't really apply. But I can't not put them in a block because that changes cspans and some assumptions in TagHelperBlockRewriter. This is probably a good candidate for cleanup later. But I don't want to change this now